File: Deque/Hard/Min_Heap_CRUD/MinHeap.py
""" Copyrights to venkys.io
For more information, visit https://venkys.io 

Space complexity: O(n)
Time complexity
- Insertion: O(log n)
- Extraction of Minimum: O(log n)
- Getting Minimum: O(1)
- Updating Key: O(n)
- Deleting Key : O(n) """

import heapq
import sys

class MinHeap:
    def __init__(self):
        """
        Initialize an empty MinHeap.

        Attributes:
        - heap: A list to store elements in heap order.
        """
        self.heap = []

    def insert(self, value):
        """
        Insert a value into the MinHeap.

        Parameters:
        - value: The value to be inserted.
        """
        heapq.heappush(self.heap, value)

    def extract_min(self):
        """
        Extract the minimum value from the MinHeap.

        Returns:
        - The minimum value if the heap is not empty, otherwise None.
        """
        if self.heap:
            return heapq.heappop(self.heap)
        return None

    def get_min(self):
        """
        Get the minimum value from the MinHeap without removing it.

        Returns:
        - The minimum value if the heap is not empty, otherwise None.
        """
        return self.heap[0] if self.heap else None

    def update_key(self, old_value, new_value):
        """
        Update a key in the MinHeap with a new value.

        Parameters:
        - old_value: The value to be updated.
        - new_value: The new value to replace the old value.
        """
        if old_value in self.heap:
            index = self.heap.index(old_value)
            self.heap[index] = new_value
            heapq.heapify(self.heap)

    def delete_key(self, value):
        """
        Delete a key from the MinHeap.

        Parameters:
        - value: The value to be deleted.
        """
        if value in self.heap:
            self.heap.remove(value)
            heapq.heapify(self.heap)

def main():
    min_heap = MinHeap()

    try:
        # Read input values from STDIN
        input_values = input("Enter space-separated values: ").split()

        # Check for null values or empty input
        if not input_values:
            raise ValueError("Error: Input values are missing.")

        # Convert input values to integers
        values = map(int, input_values)

        # Insert values into the MinHeap
        for value in values:
            min_heap.insert(value)

        print("Min Heap:", min_heap.heap)

        min_value = min_heap.extract_min()
        if min_value is not None:
            print("Extracted Min Value:", min_value)
            print("Updated Min Heap:", min_heap.heap)

        min_value = min_heap.get_min()
        if min_value is not None:
            print("Min Value:", min_value)

        # Update a key in the MinHeap
        old_value = int(input("Enter the value to be updated: "))
        new_value = int(input("Enter the new value: "))
        min_heap.update_key(old_value, new_value)
        print("Updated Min Heap:", min_heap.heap)

        # Delete a key from the MinHeap
        value_to_delete = int(input("Enter the value to be deleted: "))
        min_heap.delete_key(value_to_delete)
        print("Heap after deleting {}: {}".format(value_to_delete, min_heap.heap))

    except ValueError as e:
        print(str(e))

if __name__ == "__main__":
    main()

File: Deque/Hard/Min_Heap_CRUD/MinHeap.java
/* Copyrights to venkys.io
For more information, visit https://venkys.io */

//Space complexity: O(n)
/*Time complexity
- Insertion: O(log n)
- Extraction of Minimum: O(log n)
- Getting Minimum: O(1)
- Updating Key: O(n)
- Deleting Key : O(n)*/

import java.util.PriorityQueue;
import java.util.Scanner;

/**
 * The MinHeap class represents a minimum heap implemented using a PriorityQueue in Java.
 */
public class MinHeap {
    private PriorityQueue<Integer> heap;

    /**
     * Constructs a new MinHeap with an empty PriorityQueue.
     */
    public MinHeap() {
        this.heap = new PriorityQueue<>();
    }

    /**
     * Inserts a value into the MinHeap.
     *
     * @param value The value to be inserted.
     */
    public void insert(int value) {
        heap.add(value);
    }

    /**
     * Extracts and returns the minimum value from the MinHeap.
     *
     * @return The minimum value if the heap is not empty, otherwise null.
     */
    public Integer extractMin() {
        return heap.poll();
    }

    /**
     * Returns the minimum value from the MinHeap without removing it.
     *
     * @return The minimum value if the heap is not empty, otherwise null.
     */
    public Integer getMin() {
        return heap.peek();
    }

    /**
     * Updates a key in the MinHeap with a new value.
     *
     * @param oldValue The value to be updated.
     * @param newValue The new value to replace the old value.
     */
    public void updateKey(int oldValue, int newValue) {
        heap.remove(oldValue);
        heap.add(newValue);
    }

    /**
     * Deletes a key from the MinHeap.
     *
     * @param value The value to be deleted.
     */
    public void deleteKey(int value) {
        heap.remove(value);
    }

    /**
     * Gets the current state of the MinHeap.
     *
     * @return A string representation of the MinHeap.
     */
    public String getHeapState() {
        return heap.toString();
    }

    /**
     * The main method for demonstrating the usage of the MinHeap class.
     */
    public static void main(String[] args) {
        // Create a new MinHeap instance
        MinHeap minHeap = new MinHeap();

        try (Scanner scanner = new Scanner(System.in)) {
            // Prompt the user to enter space-separated values
            System.out.println("Enter space-separated values (or press Enter to finish):");
            String inputLine = scanner.nextLine();

            // Check if the input is null or empty
            if (inputLine == null || inputLine.trim().isEmpty()) {
                System.out.println("Error: Input values are missing or empty.");
                return;
            }

            // Split the input values and insert them into the MinHeap
            String[] inputValues = inputLine.split("\\s+");
            for (String inputValue : inputValues) {
                try {
                    int value = Integer.parseInt(inputValue);
                    minHeap.insert(value);
                } catch (NumberFormatException e) {
                    System.out.println("Error: Invalid integer value - " + inputValue);
                    return;
                }
            }

            // Display the initial Min Heap
            System.out.println("Min Heap: " + minHeap.getHeapState());

            // Extract and display the minimum value
            Integer minValue = minHeap.extractMin();
            if (minValue != null) {
                System.out.println("Extracted Min Value: " + minValue);
                System.out.println("Updated Min Heap: " + minHeap.getHeapState());
            }

            // Display the current minimum value without removing it
            Integer min = minHeap.getMin();
            if (min != null) {
                System.out.println("Min Value: " + min);
            }

            // Update a key in the MinHeap
            System.out.print("Enter the value to be updated: ");
            int oldValue;
            try {
                oldValue = scanner.nextInt();
            } catch (java.util.InputMismatchException e) {
                System.out.println("Error: Invalid input for updating value.");
                return;
            }

            System.out.print("Enter the new value: ");
            int newValue;
            try {
                newValue = scanner.nextInt();
            } catch (java.util.InputMismatchException e) {
                System.out.println("Error: Invalid input for new value.");
                return;
            }

            minHeap.updateKey(oldValue, newValue);
            System.out.println("Updated Min Heap: " + minHeap.getHeapState());

            // Delete a key from the MinHeap
            System.out.print("Enter the value to be deleted: ");
            int valueToDelete;
            try {
                valueToDelete = scanner.nextInt();
            } catch (java.util.InputMismatchException e) {
                System.out.println("Error: Invalid input for deleting value.");
                return;
            }

            minHeap.deleteKey(valueToDelete);
            System.out.println("Heap after deleting " + valueToDelete + ": " + minHeap.getHeapState());
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}